src/enzo/Grid_ActiveParticleHandler.C

```diff
@@ -0,0 +1,249 @@
+/***********************************************************************
+/
+/  GRID CLASS (HANDLE THE CREATION AND FEEDBACK OF ACTIVE PARTICLES)
+/
+/  written by: Greg Bryan
+/  date:       March, 1997
+/  modified1:  April, 2009 by JHW to have multiple types of star 
+/              particles
+/  modified2:  May, 2011 by MJT to be in support of active particles
+/
+/  PURPOSE:
+/
+/  RETURNS:
+/    SUCCESS or FAIL
+/
+************************************************************************/
+#include <assert.h> 
+#include "preincludes.h"
+#include "ErrorExceptions.h"
+#include "performance.h"
+#include "macros_and_parameters.h"
+#include "typedefs.h"
+#include "global_data.h"
+#include "Fluxes.h"
+#include "GridList.h"
+#include "ExternalBoundary.h"
+#include "Grid.h"
+#include "fortran.def"
+#include "CosmologyParameters.h"
+#include "ActiveParticle.h"
+#define H_DEBUG 0
+/* function prototypes */
+ 
+int CosmologyComputeExpansionFactor(FLOAT time, FLOAT *a, FLOAT *dadt);
+int GetUnits(float *DensityUnits, float *LengthUnits,
+	     float *TemperatureUnits, float *TimeUnits,
+	     float *VelocityUnits, FLOAT Time);
+int FindField(int field, int farray[], int numfields);
+ 
+int grid::ActiveParticleHandler(HierarchyEntry* SubgridPointer, int level,
+                                float dtLevelAbove, int &NumberOfNewParticles)
+{
+
+  if (EnabledActiveParticlesCount == 0) return SUCCESS;
+
+  if (MyProcessorNumber != ProcessorNumber)
+    return SUCCESS;
+ 
+  if (NumberOfBaryonFields == 0)
+    return SUCCESS;
+
+  /* First, set under_subgrid field */
+  HierarchyEntry *Subgrid;
+  this->ZeroSolutionUnderSubgrid(NULL, ZERO_UNDER_SUBGRID_FIELD);
+  for (Subgrid = SubgridPointer; Subgrid; Subgrid = Subgrid->NextGridThisLevel)
+    this->ZeroSolutionUnderSubgrid(Subgrid->GridData, ZERO_UNDER_SUBGRID_FIELD);
+
+  /* initialize */
+
+  LCAPERF_START("grid_ActiveParticleHandler");
+
+  /* First we identify the data dependencies */
+
+  struct ActiveParticleFormationDataFlags flags = flags_default;
+
+  int i;
```

Comment:
This loop seems strange.  Is it actually doing anything?  This is likely a shortcoming on my end, but it feels like all this does is set the values of `flags` and do nothing with it...

Reply:
Hi Dave, 
Those `flags` values are needed by the active particle type later on. For example the SmartStar particle sets the H2Fraction flag to True. This is then checked/used in `ActiveParticle.C` in the `ConstructData` function. So you can think of the `flags` structure as a booking structure that tells each AP what attributes it has. 
Does that (kinda) make sense?